File: android/app/src/main/java/com/lilru/liftr/ui/add/recommendation/WorkoutRecommendationContext.kt
package com.lilru.liftr.ui.add.recommendation

import com.lilru.liftr.data.BackendContracts
import com.lilru.liftr.ui.goals.LiftrGoalsTime
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.postgrest.postgrest
import java.time.Instant
import java.time.temporal.ChronoUnit
import kotlin.math.min
import kotlin.math.roundToInt
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

data class GoalRecommendationNudge(
    val workoutsRemaining: Int? = null,
    val caloriesRemaining: Double? = null,
    val summaryLine: String? = null
)

data class MuscleFreshnessEntryResult(
    val muscle: String,
    val status: String
)

data class StrengthRecommendationOutputResult(
    val exercises: List<StrengthRecommendationExerciseResult>,
    val sessionRationale: String? = null,
    val muscleFreshness: List<MuscleFreshnessEntryResult> = emptyList(),
    val routineName: String? = null
)

data class WorkoutRecommendationContext(
    val profileWeightKg: Double? = null,
    val profileSex: String? = null,
    val favoriteExerciseIds: Set<Long> = emptySet(),
    val prMaxWeightByExerciseId: Map<Long, Double> = emptyMap(),
    val goalNudge: GoalRecommendationNudge? = null,
    val publishedStrengthCount: Int = 0,
    val networkExerciseFrequency: Map<Long, Int> = emptyMap(),
    val partialDataNote: String? = null
) {
    val isBeginnerLifter: Boolean
        get() = publishedStrengthCount < WorkoutRecommendationConstants.BEGINNER_STRENGTH_SESSION_THRESHOLD

    fun defaultColdStartWeightKg(): Double =
        WorkoutRecommendationConstants.defaultColdStartWeightKg(profileWeightKg, isBeginnerLifter)

    fun appendGoalLine(rationale: String): String {
        val line = goalNudge?.summaryLine ?: return rationale
        if (rationale.isEmpty()) return line
        return "$rationale $line"
    }

    fun cardioDurationMultiplier(): Double {
        val rem = goalNudge?.caloriesRemaining ?: return 1.0
        if (rem <= 0) return 1.0
        return when {
            rem > 500 -> 1.15
            rem > 200 -> 1.08
            else -> 1.0
        }
    }
}

object WorkoutRecommendationConstants {
    const val LOOKBACK_COUNT = 10
    const val TARGET_EXERCISE_COUNT = 5
    const val DEFAULT_SETS_PER_EXERCISE = 3
    const val DEFAULT_REPS = 12
    const val MAX_RECOMMENDED_SETS = 5
    const val MAX_INFERRED_SETS_FROM_SET_NUMBER = 8
    const val MAX_RECOMMENDED_REPS = 22
    const val MIN_RECOMMENDED_REPS = 6
    const val HIGH_VOLUME_REPS_THRESHOLD = 17
    const val HIGH_VOLUME_SETS_THRESHOLD = 5
    const val DEFAULT_REST_BETWEEN_SETS_SEC = 90
    const val RPE_WEIGHT_DELTA_KG = 2.5
    const val RECOVERY_DEPRIORITIZE_HOURS = 48
    const val FAVORITE_SELECTION_BOOST = 2
    const val PR_CHASE_PROXIMITY_RATIO = 0.05
    const val PR_CAP_INCREMENT_KG = 2.5
    const val BEGINNER_STRENGTH_SESSION_THRESHOLD = 5
    const val NETWORK_LOOKBACK_DAYS = 7

    fun defaultColdStartWeightKg(profileWeightKg: Double?, isBeginner: Boolean): Double {
        if (profileWeightKg != null && profileWeightKg > 0) {
            val factor = if (isBeginner) 0.12 else 0.18
            return roundToHalf(min(60.0, maxOf(5.0, profileWeightKg * factor)))
        }
        return if (isBeginner) 15.0 else 20.0
    }

    fun roundToHalf(x: Double) = (x * 2.0).roundToInt() / 2.0

    fun <T> biasedShuffle(items: List<T>, isBoosted: (T) -> Boolean): List<T> {
        val boosted = mutableListOf<T>()
        val rest = mutableListOf<T>()
        for (item in items) {
            if (isBoosted(item)) {
                repeat(FAVORITE_SELECTION_BOOST) { boosted.add(item) }
            } else {
                rest.add(item)
            }
        }
        return (boosted + rest).shuffled()
    }
}

object WorkoutRecommendationContextLoader {

    @Serializable
    private data class PRRow(
        val kind: String,
        val label: String,
        val metric: String,
        val value: Double
    )

    suspend fun load(
        supabase: SupabaseClient,
        json: Json,
        userId: String,
        catalog: List<ExerciseForRecommendation>,
        includeNetwork: Boolean
    ): WorkoutRecommendationContext {
        var ctx = WorkoutRecommendationContext()

        @Serializable
        data class ProfileRow(
            @SerialName("weight_kg") val weightKg: Double? = null,
            val sex: String? = null,
            @SerialName("date_of_birth") val dateOfBirth: String? = null
        )
        runCatching {
            val pRes = supabase.from(BackendContracts.Tables.PROFILES)
                .select(Columns.raw("weight_kg, sex, date_of_birth")) {
                    filter { eq("user_id", userId) }
                }
            val rows = json.decodeFromString<List<ProfileRow>>(pRes.data)
            rows.firstOrNull()?.let { p ->
                ctx = ctx.copy(profileWeightKg = p.weightKg, profileSex = p.sex)
            }
        }

        @Serializable
        data class FavRow(@SerialName("exercise_id") val exerciseId: Long)
        runCatching {
            val fRes = supabase.from(BackendContracts.Tables.USER_FAVORITE_EXERCISES)
                .select(Columns.raw("exercise_id")) {
                    filter { eq("user_id", userId) }
                }
            val favs = json.decodeFromString<List<FavRow>>(fRes.data).map { it.exerciseId }.toSet()
            ctx = ctx.copy(favoriteExerciseIds = favs)
        }

        runCatching {
            val prParams = buildJsonObject {
                put("p_user_id", userId)
                put("p_kind", "strength")
            }
            val prRes = supabase.postgrest.rpc(BackendContracts.Rpc.GET_USER_PRS, prParams) { }
            val prs = json.decodeFromString<List<PRRow>>(prRes.data)
            ctx = ctx.copy(prMaxWeightByExerciseId = mapStrengthPrs(prs, catalog))
        }

        @Serializable
        data class CountRow(val id: Int)
        runCatching {
            val cRes = supabase.from(BackendContracts.Tables.WORKOUTS)
                .select(Columns.raw("id")) {
                    filter {
                        eq("user_id", userId)
                        eq("kind", "strength")
                        eq("state", "published")
                    }
                }
            ctx = ctx.copy(publishedStrengthCount = json.decodeFromString<List<CountRow>>(cRes.data).size)
        }

        ctx = ctx.copy(goalNudge = loadGoalNudge(supabase, json, userId))

        if (includeNetwork) {
            val (freq, note) = loadNetworkFrequency(supabase, json, userId)
            ctx = ctx.copy(networkExerciseFrequency = freq, partialDataNote = note)
        }

        return ctx
    }

    private fun mapStrengthPrs(
        prs: List<PRRow>,
        catalog: List<ExerciseForRecommendation>
    ): Map<Long, Double> {
        val metrics = setOf("weight_kg", "max_weight_kg", "weight")
        val out = mutableMapOf<Long, Double>()
        for (pr in prs) {
            if (pr.kind.lowercase() != "strength") continue
            if (pr.metric.lowercase() !in metrics) continue
            val norm = pr.label.trim().lowercase()
            if (norm.isEmpty()) continue
            for (ex in catalog) {
                val names = listOfNotNull(ex.name, ex.nameEn, ex.nameEs).map { it.trim().lowercase() }
                if (norm in names) {
                    out[ex.id] = maxOf(out[ex.id] ?: 0.0, pr.value)
                }
            }
        }
        return out
    }

    private suspend fun loadGoalNudge(
        supabase: SupabaseClient,
        json: Json,
        userId: String
    ): GoalRecommendationNudge? {
        val weekStr = LiftrGoalsTime.currentWeekStartDateString()
        @Serializable
        data class GoalRow(
            val id: Long,
            val metric: String,
            @SerialName("target_value") val targetValue: Double
        )
        @Serializable
        data class ResultRow(
            @SerialName("goal_id") val goalId: Long,
            @SerialName("achieved_value") val achievedValue: Double? = null,
            @SerialName("is_completed") val isCompleted: Boolean? = null
        )
        val gRes = runCatching {
            supabase.from(BackendContracts.Tables.WEEKLY_GOALS)
                .select(Columns.raw("id, metric, target_value")) {
                    filter {
                        eq("user_id", userId)
                        eq("week_start", weekStr)
                    }
                }
        }.getOrNull() ?: return null
        val goals = json.decodeFromString<List<GoalRow>>(gRes.data)
        if (goals.isEmpty()) return null
        val resultsByGoal = mutableMapOf<Long, ResultRow>()
        runCatching {
            val rRes = supabase.from(BackendContracts.Tables.WEEKLY_GOAL_RESULTS)
                .select(Columns.raw("goal_id, achieved_value, is_completed")) {
                    filter {
                        isIn("goal_id", goals.map { it.id.toString() })
                        eq("week_start", weekStr)
                    }
                }
            json.decodeFromString<List<ResultRow>>(rRes.data).forEach { resultsByGoal[it.goalId] = it }
        }
        val lines = mutableListOf<String>()
        var workoutsRemaining: Int? = null
        var caloriesRemaining: Double? = null
        for (g in goals) {
            val r = resultsByGoal[g.id]
            if (r?.isCompleted == true) continue
            val achieved = r?.achievedValue ?: 0.0
            when (g.metric.lowercase()) {
                "workouts" -> {
                    val rem = maxOf(0, g.targetValue.roundToInt() - achieved.roundToInt())
                    if (rem in 1..2) {
                        workoutsRemaining = rem
                        lines += if (rem == 1) {
                            "You're 1 workout away from this week's workout goal."
                        } else {
                            "You're $rem workouts away from this week's workout goal."
                        }
                    }
                }
                "calories" -> {
                    val rem = maxOf(0.0, g.targetValue - achieved)
                    if (rem > 150) {
                        caloriesRemaining = rem
                        lines += "Calorie goal still open this week—a slightly longer session can help."
                    }
                }
            }
        }
        if (lines.isEmpty()) return null
        return GoalRecommendationNudge(workoutsRemaining, caloriesRemaining, lines.joinToString(" "))
    }

    private suspend fun loadNetworkFrequency(
        supabase: SupabaseClient,
        json: Json,
        userId: String
    ): Pair<Map<Long, Int>, String?> {
        @Serializable
        data class FollowRow(@SerialName("followee_id") val followeeId: String)
        val fRes = runCatching {
            supabase.from(BackendContracts.Tables.FOLLOWS)
                .select(Columns.raw("followee_id")) {
                    filter { eq("follower_id", userId) }
                    limit(50)
                }
        }.getOrNull() ?: return emptyMap<Long, Int>() to null
        val followees = json.decodeFromString<List<FollowRow>>(fRes.data).map { it.followeeId }
        if (followees.isEmpty()) return emptyMap<Long, Int>() to null
        val since = Instant.now().minus(WorkoutRecommendationConstants.NETWORK_LOOKBACK_DAYS.toLong(), ChronoUnit.DAYS)
        @Serializable
        data class WRow(val id: Int)
        val wRes = runCatching {
            supabase.from(BackendContracts.Tables.WORKOUTS)
                .select(Columns.raw("id")) {
                    filter {
                        isIn("user_id", followees)
                        eq("kind", "strength")
                        eq("state", "published")
                        gte("started_at", since.toString())
                    }
                    limit(100)
                }
        }.getOrNull() ?: return emptyMap<Long, Int>() to "Your network hasn't logged strength workouts recently—we used your own history."
        val workouts = json.decodeFromString<List<WRow>>(wRes.data)
        if (workouts.isEmpty()) {
            return emptyMap<Long, Int>() to "Your network hasn't logged strength workouts recently—we used your own history."
        }
        @Serializable
        data class ExRow(@SerialName("exercise_id") val exerciseId: Long)
        val exRes = runCatching {
            supabase.from(BackendContracts.Tables.WORKOUT_EXERCISES)
                .select(Columns.raw("exercise_id")) {
                    filter { isIn("workout_id", workouts.map { it.id.toString() }) }
                }
        }.getOrNull() ?: return emptyMap<Long, Int>() to "Could not load network exercise trends."
        val freq = json.decodeFromString<List<ExRow>>(exRes.data)
            .groupingBy { it.exerciseId }.eachCount()
        return freq to null
    }
}

File: Liftr/WorkoutRecommendationContext.swift
import Foundation
import Supabase

private struct StrengthPRRow: Decodable {
    let kind: String
    let label: String
    let metric: String
    let value: Double
}

struct GoalRecommendationNudge: Equatable {
    let workoutsRemaining: Int?
    let caloriesRemaining: Double?
    let summaryLine: String?
}

struct MuscleFreshnessEntry: Identifiable, Equatable {
    enum Status: String, Equatable {
        case fresh
        case recent
        case trainedRecently
    }
    let id: String
    let muscle: String
    let status: Status
}

struct StrengthRecommendationOutput: Equatable {
    let exercises: [StrengthRecommendationExercise]
    let sessionRationale: String?
    let muscleFreshness: [MuscleFreshnessEntry]
    let routineName: String?
}

struct WorkoutRecommendationContext: Equatable {
    var profileWeightKg: Double?
    var profileSex: String?
    var profileAgeYears: Int?
    var favoriteExerciseIds: Set<Int64> = []
    var prMaxWeightByExerciseId: [Int64: Double] = [:]
    var goalNudge: GoalRecommendationNudge?
    var publishedStrengthCount: Int = 0
    var networkExerciseFrequency: [Int64: Int] = [:]
    var partialDataNote: String?

    var isBeginnerLifter: Bool { publishedStrengthCount < WorkoutRecommendationConstants.beginnerStrengthSessionThreshold }

    static func load(
        userId: UUID,
        catalog: [Exercise],
        includeNetwork: Bool
    ) async -> WorkoutRecommendationContext {
        let client = SupabaseManager.shared.client
        let decoder = JSONDecoder.supabase()
        var ctx = WorkoutRecommendationContext()

        struct ProfileRow: Decodable {
            let weight_kg: Double?
            let sex: String?
            let date_of_birth: String?
        }
        if let pRes = try? await client
            .from("profiles")
            .select("weight_kg, sex, date_of_birth")
            .eq("user_id", value: userId.uuidString)
            .single()
            .execute(),
           let profile = try? decoder.decode(ProfileRow.self, from: pRes.data) {
            ctx.profileWeightKg = profile.weight_kg
            ctx.profileSex = profile.sex
            if let dob = profile.date_of_birth {
                let df = DateFormatter()
                df.locale = Locale(identifier: "en_US_POSIX")
                df.dateFormat = "yyyy-MM-dd"
                if let birth = df.date(from: dob) {
                    ctx.profileAgeYears = Calendar.current.dateComponents([.year], from: birth, to: Date()).year
                }
            }
        }

        struct FavRow: Decodable { let exercise_id: Int64 }
        if let fRes = try? await client.from("user_favorite_exercises").select("exercise_id").eq("user_id", value: userId.uuidString).execute(),
           let rows = try? decoder.decode([FavRow].self, from: fRes.data) {
            ctx.favoriteExerciseIds = Set(rows.map(\.exercise_id))
        }

        if let prRes = try? await client.rpc("get_user_prs", params: ["p_user_id": userId.uuidString, "p_kind": "strength"]).execute(),
           let prs = try? decoder.decode([StrengthPRRow].self, from: prRes.data) {
            ctx.prMaxWeightByExerciseId = mapStrengthPrsToExerciseIds(prs: prs, catalog: catalog)
        }

        struct CountRow: Decodable { let id: Int }
        if let cRes = try? await client
            .from("workouts")
            .select("id")
            .eq("user_id", value: userId.uuidString)
            .eq("kind", value: "strength")
            .eq("state", value: "published")
            .execute(),
           let counts = try? decoder.decode([CountRow].self, from: cRes.data) {
            ctx.publishedStrengthCount = counts.count
        }

        ctx.goalNudge = await loadGoalNudge(userId: userId, client: client, decoder: decoder)

        if includeNetwork {
            let (freq, note) = await loadNetworkExerciseFrequency(userId: userId, client: client, decoder: decoder)
            ctx.networkExerciseFrequency = freq
            if let note { ctx.partialDataNote = note }
        }

        return ctx
    }

    func defaultColdStartWeightKg() -> Double {
        WorkoutRecommendationConstants.defaultColdStartWeightKg(
            profileWeightKg: profileWeightKg,
            isBeginner: isBeginnerLifter
        )
    }

    func appendGoalLine(to rationale: String) -> String {
        guard let line = goalNudge?.summaryLine, !line.isEmpty else { return rationale }
        if rationale.isEmpty { return line }
        return rationale + " " + line
    }

    func cardioDurationMultiplier() -> Double {
        guard let rem = goalNudge?.caloriesRemaining, rem > 0 else { return 1.0 }
        if rem > 500 { return 1.15 }
        if rem > 200 { return 1.08 }
        return 1.0
    }

    private static func mapStrengthPrsToExerciseIds(prs: [StrengthPRRow], catalog: [Exercise]) -> [Int64: Double] {
        var out: [Int64: Double] = [:]
        let weightMetrics = Set(["weight_kg", "max_weight_kg", "weight"])
        for pr in prs where pr.kind.lowercased() == "strength" && weightMetrics.contains(pr.metric.lowercased()) {
            let normLabel = pr.label.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines).lowercased()
            guard !normLabel.isEmpty else { continue }
            for ex in catalog {
                let names = ([ex.name] + [ex.name_en, ex.name_es].compactMap { $0 })
                    .map { $0.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines).lowercased() }
                if names.contains(normLabel) {
                    out[ex.id] = max(out[ex.id] ?? 0, pr.value)
                }
            }
        }
        return out
    }

    private static func loadGoalNudge(userId: UUID, client: SupabaseClient, decoder: JSONDecoder) async -> GoalRecommendationNudge? {
        let weekStr = GoalsManager.dateOnlyString(GoalsManager.currentWeekStart())
        struct GoalRow: Decodable {
            let id: Int64
            let metric: String
            let target_value: Double
        }
        struct ResultRow: Decodable {
            let goal_id: Int64
            let achieved_value: Double?
            let is_completed: Bool?
        }
        guard let gRes = try? await client
            .from("weekly_goals")
            .select("id, metric, target_value")
            .eq("user_id", value: userId.uuidString)
            .eq("week_start", value: weekStr)
            .execute(),
              let goals = try? decoder.decode([GoalRow].self, from: gRes.data),
              !goals.isEmpty else { return nil }

        let ids = goals.map { String($0.id) }
        var resultsByGoal: [Int64: ResultRow] = [:]
        if let rRes = try? await client
            .from("weekly_goal_results")
            .select("goal_id, achieved_value, is_completed")
            .in("goal_id", values: ids)
            .eq("week_start", value: weekStr)
            .execute(),
           let results = try? decoder.decode([ResultRow].self, from: rRes.data) {
            for r in results { resultsByGoal[r.goal_id] = r }
        }

        var workoutsRemaining: Int?
        var caloriesRemaining: Double?
        var lines: [String] = []

        for g in goals {
            let achieved = resultsByGoal[g.id]?.achieved_value ?? 0
            let done = resultsByGoal[g.id]?.is_completed ?? false
            if done { continue }
            switch g.metric.lowercased() {
            case "workouts":
                let rem = max(0, Int(g.target_value.rounded()) - Int(achieved.rounded()))
                if rem > 0 && rem <= 2 {
                    workoutsRemaining = rem
                    lines.append(rem == 1 ? "You're 1 workout away from this week's workout goal." : "You're \(rem) workouts away from this week's workout goal.")
                }
            case "calories":
                let rem = max(0, g.target_value - achieved)
                if rem > 150 {
                    caloriesRemaining = rem
                    lines.append("Calorie goal still open this week—a slightly longer session can help.")
                }
            default:
                break
            }
        }

        guard !lines.isEmpty else { return nil }
        return GoalRecommendationNudge(
            workoutsRemaining: workoutsRemaining,
            caloriesRemaining: caloriesRemaining,
            summaryLine: lines.joined(separator: " ")
        )
    }

    private static func loadNetworkExerciseFrequency(
        userId: UUID,
        client: SupabaseClient,
        decoder: JSONDecoder
    ) async -> ([Int64: Int], String?) {
        struct FollowRow: Decodable { let followee_id: UUID }
        guard let fRes = try? await client
            .from("follows")
            .select("followee_id")
            .eq("follower_id", value: userId.uuidString)
            .limit(50)
            .execute(),
              let follows = try? decoder.decode([FollowRow].self, from: fRes.data),
              !follows.isEmpty else { return ([:], nil) }

        let followeeIds = follows.map { $0.followee_id.uuidString }
        let since = Calendar.current.date(byAdding: .day, value: -WorkoutRecommendationConstants.networkLookbackDays, to: Date()) ?? Date()
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime]

        struct WRow: Decodable { let id: Int }
        guard let wRes = try? await client
            .from("workouts")
            .select("id")
            .in("user_id", values: followeeIds)
            .eq("kind", value: "strength")
            .eq("state", value: "published")
            .gte("started_at", value: iso.string(from: since))
            .limit(100)
            .execute(),
              let workouts = try? decoder.decode([WRow].self, from: wRes.data),
              !workouts.isEmpty else {
            return ([:], "Your network hasn't logged strength workouts recently—we used your own history.")
        }

        let wids = workouts.map { String($0.id) }
        struct ExRow: Decodable { let exercise_id: Int64 }
        guard let exRes = try? await client
            .from("workout_exercises")
            .select("exercise_id")
            .in("workout_id", values: wids)
            .execute(),
              let exRows = try? decoder.decode([ExRow].self, from: exRes.data) else {
            return ([:], "Could not load network exercise trends.")
        }

        var freq: [Int64: Int] = [:]
        for row in exRows { freq[row.exercise_id, default: 0] += 1 }
        return (freq, nil)
    }
}

enum WorkoutRecommendationConstants {
    static let lookbackCount = 10
    static let targetExerciseCount = 5
    static let defaultSetsPerExercise = 3
    static let defaultReps = 12
    static let maxRecommendedSets = 5
    static let maxInferredSetsFromSetNumber = 8
    static let maxRecommendedReps = 22
    static let minRecommendedReps = 6
    static let highVolumeRepsThreshold = 17
    static let highVolumeSetsThreshold = 5
    static let defaultRestBetweenSetsSec = 90
    static let rpeWeightDeltaKg = 2.5
    static let recoveryDeprioritizeHours = 48
    static let favoriteSelectionBoost = 2
    static let prChaseProximityRatio = 0.05
    static let prCapIncrementKg = 2.5
    static let beginnerStrengthSessionThreshold = 5
    static let networkLookbackDays = 7

    static func defaultColdStartWeightKg(profileWeightKg: Double?, isBeginner: Bool) -> Double {
        if let w = profileWeightKg, w > 0 {
            let factor = isBeginner ? 0.12 : 0.18
            let raw = w * factor
            return roundToHalf(min(60, max(5, raw)))
        }
        return isBeginner ? 15 : 20
    }

    static func roundToHalf(_ x: Double) -> Double {
        (x * 2).rounded() / 2
    }

    static func biasedShuffle<T>(_ items: [T], isBoosted: (T) -> Bool) -> [T] {
        var boosted: [T] = []
        var rest: [T] = []
        for item in items {
            if isBoosted(item) {
                for _ in 0..<favoriteSelectionBoost { boosted.append(item) }
            } else {
                rest.append(item)
            }
        }
        return (boosted + rest).shuffled()
    }
}

func strengthRecommendationsFromRoutineDetail(
    _ detail: StrengthTemplateDetailWire,
    catalog: [Exercise],
    exerciseLanguage: ExerciseLanguage
) -> [StrengthRecommendationExercise] {
    let muscleById = Dictionary(uniqueKeysWithValues: catalog.map { ($0.id, $0.muscle_primary) })
    let exs = (detail.strength_routine_exercises ?? []).sorted { $0.order_index < $1.order_index }
    return exs.compactMap { ex -> StrengthRecommendationExercise? in
        let setsSorted = (ex.strength_routine_sets ?? []).sorted { $0.set_number < $1.set_number }
        let setsOut: [StrengthRecommendationSet] = setsSorted.enumerated().map { idx, s in
            StrengthRecommendationSet(
                setNumber: idx + 1,
                reps: max(WorkoutRecommendationConstants.minRecommendedReps, s.reps ?? WorkoutRecommendationConstants.defaultReps),
                weightKg: WorkoutRecommendationConstants.roundToHalf(s.weight_kg ?? 0),
                rpe: s.rpe,
                restSec: s.rest_sec ?? WorkoutRecommendationConstants.defaultRestBetweenSetsSec
            )
        }
        let fallbackSets = setsOut.isEmpty
            ? (1...WorkoutRecommendationConstants.defaultSetsPerExercise).map { sn in
                StrengthRecommendationSet(
                    setNumber: sn,
                    reps: WorkoutRecommendationConstants.defaultReps,
                    weightKg: 20,
                    rpe: 8,
                    restSec: WorkoutRecommendationConstants.defaultRestBetweenSetsSec
                )
            }
            : setsOut
        let displayName: String = {
            if let cn = ex.custom_name?.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines), !cn.isEmpty { return cn }
            if let cat = catalog.first(where: { $0.id == ex.exercise_id }) {
                return cat.localizedName(for: exerciseLanguage)
            }
            return "Exercise \(ex.exercise_id)"
        }()
        return StrengthRecommendationExercise(
            exerciseId: ex.exercise_id,
            displayName: displayName,
            musclePrimary: muscleById[ex.exercise_id].flatMap { $0 },
            sets: fallbackSets
        )
    }
}
